### PR TITLE
Add slugs to URLs to make them friendlier

### LINKED ISSUE #15256: "Duplicate" dashboard modal says it "failed", when it actually did duplicate successfully
**Describe the bug**
When users with curate access wants to duplicate a dashboard in a collection they are presented with the modal where they can enter new name, description, choose collection and finally click "Duplicate" button. Once that button is clicked, it turns to red-danger button "FAILED", but at the same time you can see bubble with the "Dashboard duplicated - See it" in the lower left corner.

Note, dashboard is indeed duplicated.

**Logs**
No errors.

**To Reproduce**
_As an admin or normal user with collection curate access level_
1. Go to `/collection/root`
2. Click on `...` menu for any dashboard in the collection
3. Choose "Duplicate this item"
4. Modal appears, click "Duplicate" button
5. Notice it says FAILED
    - Notice the bubble with the "Dashboard duplicated - See it" in the lower left corner.
    - Notice the dashboard being duplicated underneath the modal that is still open, btw
6. You can click on the "Duplicate" button in the open modal as many times as you want and it will keep duplicating dashboards ad infinitum
    - Weird thing: you can duplicate the dashboard even when the button says "FAILED"

**Expected behavior**
If users have curate permissions, they should be able to duplicate the dashboard by clicking on the "Duplicate" button in the modal that should then properly close.

**Screenshots**

**Information about your Metabase Installation:**
- Metabase version: v0.38.3-SNAPSHOT
- Metabase hosting environment: local dev, release-x.38.x
- Metabase internal database: H2 (default), sample dataset

**Severity**
P3?

**Additional context**
If you click on "See it" link in the bubble, it will properly take you to the duplicated dashboard.

### LINKED ISSUE #15765: Add slugs to URLs to make them friendlier
# Context

URLs aren't the most human-friendly thing in the world, but they are the way to point to things on the web. They're also how people share Metabase entities with each other, like:

1. `/dashboard/330/`
2. `/question/274`
3. `/browse/13`
4. `/question#<500-base-64-encoded-characters`

We can't do much about 4, as it embodies a lot more information than just an id, but we can make 1, 2 and 3 more understandable when shared. This is a little project to do that.

# What

Let's append friendly names, as slugs, to links to named entities within Metabase. It's a small thing, but it can help people understand, reference and consume content within the app, and hopefully increase its usage. It's also just nice.

### Logic

This works by keeping the URL the same, and just adding a "slugarized" version of the title of the entity at end of the URL. When parsing the URL for routing within the app, we drop the slug and only take the id into account, as we do today.

By slugarizing, I mean doing what the [pelle/slugger](https://github.com/pelle/slugger) library does, for example. (While I'm not prescribing that we use the library, it does seem to do a really good job of handling more exotic scenarios beyond the gist below).

1. Convert accented latin characters to their ASCII equivalents
2. Convert non-latin UTF-8 characters to empty chars (or some smart equivalent, not strictly needed)
3. Lowercase the string
4. Replace punctuation with spaces
5. Trim excess whitespace around and within the string
6. Replace remaining spaces with dashes

So that the three first URLs above become:

1. `/dashboard/330-pricing-v2-traction`
2. `/question/274-queries-run-in-the-last-24-weeks`
3. `/browse/33-metabase-ga`

### Scope

- Questions
- Dashboards
- Databases

### PR DESCRIPTION
**Pull request by kulyk**
This PR appends friendly entity name slugs to URLs. It can help people understand, reference, and consume content within the application. Updates links to databases, questions, dashboards, and collections. Admin links are out of scope here.

**Examples**

* `/question/1` would now be `/question/1-avg-orders-per-month`
* `/dashboard/33` would now be `/dashboard/33-customer-success-metrics`
* `/browse/4` would now be `/browse/4-app-db`
* `/collection/3` would now be `/collection/3-engineering-team`

**Note on IDs and slugs**

When parsing a URL, we just drop the slug and only use the ID, which means:

1. URLs without slugs will still work fine
2. People can just put anything instead of a slug, and it will work if an entity ID is valid. So if a question with ID 1 is actually called "Monthly Orders", but we try to visit `/question/1-whatever`, it will resolve to "Monthly Orders".

**Note about collection URLs**

* "Our analytics" is still `/collection/root`
* personal collection contains user's name (`/collection/1-bobby-tables-s-personal-collection`)
* other users' personal collections are also slugified (`collection/7-read-only-tableton-s-personal-collection`)

### To Verify

Just jump between any questions, dashboards, databases, or collections.

Closes #15765 
Closes #15256

### CODE REVIEW

**Comment by kulyk** [comment]
One more thing. This PR also fixes a few small bugs I found while working on URLs

1. When opening the "Saved Questions" browse page (`/browse/-1337`), "Saved Questions" was missing in breadcrumbs
2. When duplicating a dashboard from the Collections page, a modal was not closing automatically as it usually does

File: frontend/src/metabase/collections/components/BulkActions.jsx [hunk 1]
```diff
@@ -98,8 +98,7 @@ export default function BulkActions(props) {
             entityObject={selectedItems[0]}
             onClose={handleCloseModal}
             onSaved={newEntityObject => {
-              this.handleCloseModal();
-              this.handleBulkActionSuccess();
+              handleCloseModal();
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/collections/components/BulkActions.jsx:101]: In that case, the repro should've been unskipped together with this commit, and there should be a clear reference to this issue in the PR description.

**Review by nemanjaglumac — CHANGES_REQUESTED** [review]
Left a few comments that need to be addressed.
Summary:
1. Please unskip related repros (for the small issues that you fixed along the way) and clearly reference them in the PR description
2. Don't create any new folders inside `frontend/test/metabase/scenarios`


Note: I have reviewed only test files so far.

**Review by flamber — COMMENTED** [review]
Seems to have problems with symbol-based languages:
Danish: `æbleflæsk` => `/dashboard/1-aebleflaesk`
Chinese: `培根` => `/dashboard/1-`
Russian: `бекон` => `/dashboard/1-bekon`
Japanese: `ベーコン` => `/dashboard/1-`

Though it's not doing any magic in `/reference` or the x-rays (probably less important) `/auto/...`

And interestingly it's sluggifying certain symbols - try giving a name like `$ & % | # " \\` => `/dashboard/1-dollar-and-percent-or-`

File: frontend/src/metabase/collections/components/BulkActions.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/collections/components/BulkActions.jsx:101]: Thanks! I turned on the test locally, but it fails for non-admin users. After the dashboard is duplicated, we assert the duplicated records is present in "Our analytics". However, the duplicate popup initially suggests saving a new dashboard to the user's personal collection. This behavior is fixed by #15999. I'll rebase the PR on `master` as soon as #15999 is merged and commit the unskipped repro

**Review by kdoh — COMMENTED** [review]
Overall looking good in my testing around the basic cases, although we'll want to deal with the internationalization situation @flamber found around symbols. 

Just for fun I created a few collections with Emoji only names and in every case it's appending `%3F%3F` to the id for the slug (e.x. `7-%3F%3F`). Doesn't seem to cause any issues (and it's not that important in the grand scheme of things) but just found it curious.

**Comment by brunobergher** [comment]
This approach makes a ton of sense. It may not encompass all cases but this is is an optional/cosmetic feature in that links continue to be shareable, just carry less info. Also given the design, we can always iterate on this algorithm, since the slug doesn't affect the resolution of the URL.

👍 👍

**Review by alxnddr — COMMENTED** [review]
Overall looks great! 

I agree that symbols like `$` should be just removed instead of extrapolating.

Also, I noticed that we don't have it on the reference pages, is this intentional?
Screen Shot 2021-05-11 at 19 27 49

And one more modal right after adding a database also lacks a slug
Screen Shot 2021-05-11 at 19 23 11

**Review by brunobergher — APPROVED** [review]

**Comment by kulyk** [comment]
@alxnddr the modal you mentioned is on `/admin` right? If so, admin URLs are out of scope here. The main goal of this project is to make it easier to reference and understand the content within the app. Admin URLs are not shared too often, so adding slugs here is not that necessary. Kudos for checking everything so accurately!

117851852-65a03e00-b28f-11eb-9835-4b291638a4dc

**Comment by kulyk** [comment]
@kdoh @flamber fixed slugs for symbol-based languages and edge cases like emojis, characters, etc.

@nemanjaglumac enabled the #15256 issue reference and linked at within the PR (via "Closes" keyword), also moved the new `urls` Cypress tests to `onboarding` group as you asked

Huge thanks for reviewing such a massive PR 🙏

**Review by alxnddr — APPROVED** [review]
Looks great 🥇

File: frontend/test/metabase/scenarios/collections/permissions.cy.spec.js [hunk 2]
```diff
@@ -116,7 +116,7 @@ describe("collection permissions", () => {
               });
 
               describe("duplicate", () => {
-                it.skip("should be able to duplicate the dashboard without obstructions from the modal (metabase#15255)", () => {
+                it("should be able to duplicate the dashboard without obstructions from the modal (metabase#15256)", () => {
```
Comment (nemanjaglumac) [inline frontend/test/metabase/scenarios/collections/permissions.cy.spec.js:119]: Thank you for catching and fixing this!

**Review by nemanjaglumac — APPROVED** [review]
Looks good! 🎉

**Comment by JayJayM** [comment]
Question for you @kulyk:
What happens if the name of the dashboard changes? Would the link break? 

Our users often bookmark dashboards URLs. We don't want to break that link if we change the name of a dashboard.

cc @brunobergher

**Comment by kulyk** [comment]
Hi @JayJayM. If you change a question / dashboard / collection name, all their previous URLs will still be valid. As long as there is a correct id before a slug, the links should work

**Pull request merged**